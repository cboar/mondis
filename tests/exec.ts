import { mondisTest, oid } from './setup';

describe('Exec CachedQueries', () => {
  mondisTest('config:{} exec:{}', async (ctx) => {
    await ctx.execAll({});
    await ctx.expectRedisSnapshot('Q:*');
  });

  mondisTest('config:{} exec:{ skip: 5 }', async (ctx) => {
    await ctx.execAll({ skip: 5 });
  });

  mondisTest('config:{} exec:{ limit: 3 }', async (ctx) => {
    await ctx.execAll({ limit: 3 });
  });

  mondisTest('config:{} exec:{ skip: 5, limit: 3 }', async (ctx) => {
    await ctx.execAll({ skip: 5, limit: 3 });
  });

  mondisTest('config:{ cacheCount: 3 } exec:{}', async (ctx) => {
    await ctx.execAll({});
    await ctx.expectRedisSnapshot('Q:*');
  }, { cacheCount: 3 });

  mondisTest('config:{ cacheCount: 3 } exec:{ skip: 5 }', async (ctx) => {
    await ctx.execAll({ skip: 5 });
  }, { cacheCount: 3 });

  mondisTest('config:{ cacheCount: 3 } exec:{ limit: 3 }', async (ctx) => {
    await ctx.execAll({ limit: 3 });
    await new Promise((res) => {
      setTimeout(res, 500);
    });
    await ctx.expectRedisSnapshot('Q:*');
  }, { cacheCount: 3 });

  mondisTest('config:{ cacheCount: 3 } exec:{ skip: 5, limit: 3 }', async (ctx) => {
    await ctx.execAll({ skip: 5, limit: 3 });
  }, { cacheCount: 3 });

  mondisTest('Exec all twice, config:{}', async (ctx) => {
    await ctx.execAll({});
    await ctx.execAll({});
  });

  mondisTest('Exec all twice, config:{ cacheCount: 3 }', async (ctx) => {
    await ctx.execAll({});
    await ctx.execAll({});
  }, { cacheCount: 3 });

  mondisTest('Only count queries', async (ctx) => {
    const { queries: q } = ctx.mondis;
    await Promise.all([
      q.Static1.count(),
      q.Static2.count(),
      q.Dynamic1.count({ params: ['truck'] }),
      q.Dynamic2.count({ params: ['A'] }),
      q.Dynamic3.count({ params: [oid('D2')] }),
      q.Complex1.count({ params: [['car', 'bus']] }),
      q.Complex2.count({ params: [5000] }),
      q.Unique1.count([oid('A4')]),
      q.Populated1.count(),
      q.Populated2.count(),
      q.Sorted1.count(),
      q.Targeted1.count({ params: [[oid('B1'), oid('B2')]] }),
    ]);
    await ctx.expectRedisSnapshot('Q:*');
  });

  mondisTest('Exec filtered query', async ({ mondis }) => {
    const { Static2, Dynamic1 } = mondis.queries;
    await Static2.execWithCount({ filter: (doc) => (doc.price > 2500) });
    await Dynamic1.count({
      params: ['car'],
      filter: (doc) => (doc.price > 2500),
    });
  });

  mondisTest('Calculate hash value', ({ mondis, hashMocks }) => {
    const { Static1, Dynamic1 } = mondis.queries;
    hashMocks.forEach((spy) => spy.mockRestore());
    expect(Static1.hash).toBeTruthy();
    expect(Dynamic1.hash).toBeTruthy();
  });
});
