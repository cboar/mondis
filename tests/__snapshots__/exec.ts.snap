// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exec CachedQueries Only count queries 1`] = `
{
  "Q:Complex1________[["car","bus"]]": {
    "count": 7,
    "data": null,
  },
  "Q:Complex2________[5000]": {
    "count": 5,
    "data": null,
  },
  "Q:Dynamic1________["truck"]": {
    "count": 4,
    "data": null,
  },
  "Q:Dynamic2________["A"]": {
    "count": 5,
    "data": null,
  },
  "Q:Dynamic3________[{"$oid":"0000000000000000000000d2"}]": {
    "count": 0,
    "data": null,
  },
  "Q:Populated1______[]": {
    "count": 11,
    "data": null,
  },
  "Q:Populated2______[]": {
    "count": 11,
    "data": null,
  },
  "Q:Sorted1_________[]": {
    "count": 5,
    "data": null,
  },
  "Q:Static1_________[]": {
    "count": 4,
    "data": null,
  },
  "Q:Static2_________[]": {
    "count": 5,
    "data": null,
  },
  "Q:Targeted1_______[[{"$oid":"0000000000000000000000b1"},{"$oid":"0000000000000000000000b2"}]]": {
    "count": 2,
    "data": null,
  },
  "Q:Unique1_________[{"$oid":"0000000000000000000000a4"}]": {
    "count": 1,
    "data": null,
  },
}
`;

exports[`Exec CachedQueries config:{ cacheCount: 3 } exec:{ limit: 3 } 1`] = `
{
  "Q:Complex1________[["car","bus"]]": {
    "count": 7,
    "data": [
      {
        "_id": "0000000000000000000000a1",
        "kind": "car",
      },
      {
        "_id": "0000000000000000000000a2",
        "kind": "car",
      },
      {
        "_id": "0000000000000000000000a3",
        "kind": "car",
      },
    ],
  },
  "Q:Complex2________[5000]": {
    "count": 5,
    "data": [
      {
        "_id": "0000000000000000000000b3",
        "price": 8000,
      },
      {
        "_id": "0000000000000000000000b4",
        "price": 10000,
      },
      {
        "_id": "0000000000000000000000c1",
        "price": 20000,
      },
    ],
  },
  "Q:Dynamic1________["truck"]": {
    "count": 4,
    "data": [
      {
        "_id": "0000000000000000000000b1",
        "price": 2000,
      },
      {
        "_id": "0000000000000000000000b2",
        "price": 4000,
      },
      {
        "_id": "0000000000000000000000b3",
        "price": 8000,
      },
    ],
  },
  "Q:Dynamic2________["A"]": {
    "count": 5,
    "data": [
      {
        "_id": "0000000000000000000000a1",
        "routes": [
          "A",
          "B",
        ],
      },
      {
        "_id": "0000000000000000000000a2",
        "routes": [
          "A",
          "C",
        ],
      },
      {
        "_id": "0000000000000000000000b2",
        "routes": [
          "A",
          "D",
        ],
      },
    ],
  },
  "Q:Dynamic3________[{"$oid":"0000000000000000000000d2"}]": {
    "count": 0,
    "data": [],
  },
  "Q:Populated1______[]": {
    "count": 11,
    "data": [
      {
        "_id": "0000000000000000000000a1",
      },
      {
        "_id": "0000000000000000000000a2",
        "driver": {
          "__v": 0,
          "_id": "0000000000000000000000d1",
          "name": "Henry",
          "salary": 55000,
        },
      },
      {
        "_id": "0000000000000000000000a3",
        "driver": {
          "__v": 0,
          "_id": "0000000000000000000000d2",
          "name": "John",
          "salary": 61000,
        },
      },
    ],
  },
  "Q:Populated2______[]": {
    "count": 11,
    "data": [
      {
        "_id": "0000000000000000000000a1",
      },
      {
        "_id": "0000000000000000000000a2",
        "driver": {
          "_id": "0000000000000000000000d1",
          "name": "Henry",
        },
      },
      {
        "_id": "0000000000000000000000a3",
        "driver": {
          "_id": "0000000000000000000000d2",
          "name": "John",
        },
      },
    ],
  },
  "Q:Sorted1_________[]": {
    "count": 5,
    "data": [
      {
        "_id": "0000000000000000000000b3",
      },
      {
        "_id": "0000000000000000000000b4",
      },
      {
        "_id": "0000000000000000000000c1",
      },
    ],
  },
  "Q:Static1_________[]": {
    "count": 4,
    "data": [
      {
        "_id": "0000000000000000000000a1",
      },
      {
        "_id": "0000000000000000000000a2",
      },
      {
        "_id": "0000000000000000000000a3",
      },
    ],
  },
  "Q:Static2_________[]": {
    "count": 5,
    "data": [
      {
        "_id": "0000000000000000000000b3",
        "price": 8000,
      },
      {
        "_id": "0000000000000000000000b4",
        "price": 10000,
      },
      {
        "_id": "0000000000000000000000c1",
        "price": 20000,
      },
    ],
  },
  "Q:Targeted1_______[[{"$oid":"0000000000000000000000b1"},{"$oid":"0000000000000000000000b2"}]]": {
    "count": 2,
    "data": [
      {
        "__v": 0,
        "_id": "0000000000000000000000b1",
        "kind": "truck",
        "price": 2000,
        "routes": [],
      },
      {
        "__v": 0,
        "_id": "0000000000000000000000b2",
        "kind": "truck",
        "price": 4000,
        "routes": [
          "A",
          "D",
        ],
      },
    ],
  },
  "Q:Unique1_________[{"$oid":"0000000000000000000000a4"}]": {
    "count": null,
    "data": [
      {
        "__v": 0,
        "_id": "0000000000000000000000a4",
        "driver": "0000000000000000000000d3",
        "kind": "car",
        "price": 4000,
        "routes": [
          "B",
          "C",
        ],
      },
    ],
  },
}
`;

exports[`Exec CachedQueries config:{ cacheCount: 3 } exec:{} 1`] = `
{
  "Q:Complex1________[["car","bus"]]": {
    "count": 7,
    "data": null,
  },
  "Q:Complex2________[5000]": {
    "count": 5,
    "data": null,
  },
  "Q:Dynamic1________["truck"]": {
    "count": 4,
    "data": null,
  },
  "Q:Dynamic2________["A"]": {
    "count": 5,
    "data": null,
  },
  "Q:Dynamic3________[{"$oid":"0000000000000000000000d2"}]": {
    "count": 0,
    "data": null,
  },
  "Q:Populated1______[]": {
    "count": 11,
    "data": null,
  },
  "Q:Populated2______[]": {
    "count": 11,
    "data": null,
  },
  "Q:Sorted1_________[]": {
    "count": 5,
    "data": null,
  },
  "Q:Static1_________[]": {
    "count": 4,
    "data": null,
  },
  "Q:Static2_________[]": {
    "count": 5,
    "data": null,
  },
  "Q:Targeted1_______[[{"$oid":"0000000000000000000000b1"},{"$oid":"0000000000000000000000b2"}]]": {
    "count": 2,
    "data": null,
  },
  "Q:Unique1_________[{"$oid":"0000000000000000000000a4"}]": {
    "count": null,
    "data": [
      {
        "__v": 0,
        "_id": "0000000000000000000000a4",
        "driver": "0000000000000000000000d3",
        "kind": "car",
        "price": 4000,
        "routes": [
          "B",
          "C",
        ],
      },
    ],
  },
}
`;

exports[`Exec CachedQueries config:{} exec:{} 1`] = `
{
  "Q:Complex1________[["car","bus"]]": {
    "count": 7,
    "data": [
      {
        "_id": "0000000000000000000000a1",
        "kind": "car",
      },
      {
        "_id": "0000000000000000000000a2",
        "kind": "car",
      },
      {
        "_id": "0000000000000000000000a3",
        "kind": "car",
      },
      {
        "_id": "0000000000000000000000a4",
        "kind": "car",
      },
      {
        "_id": "0000000000000000000000c1",
        "kind": "bus",
      },
      {
        "_id": "0000000000000000000000c2",
        "kind": "bus",
      },
      {
        "_id": "0000000000000000000000c3",
        "kind": "bus",
      },
    ],
  },
  "Q:Complex2________[5000]": {
    "count": 5,
    "data": [
      {
        "_id": "0000000000000000000000b3",
        "price": 8000,
      },
      {
        "_id": "0000000000000000000000b4",
        "price": 10000,
      },
      {
        "_id": "0000000000000000000000c1",
        "price": 20000,
      },
      {
        "_id": "0000000000000000000000c2",
        "price": 40000,
      },
      {
        "_id": "0000000000000000000000c3",
        "price": 60000,
      },
    ],
  },
  "Q:Dynamic1________["truck"]": {
    "count": 4,
    "data": [
      {
        "_id": "0000000000000000000000b1",
        "price": 2000,
      },
      {
        "_id": "0000000000000000000000b2",
        "price": 4000,
      },
      {
        "_id": "0000000000000000000000b3",
        "price": 8000,
      },
      {
        "_id": "0000000000000000000000b4",
        "price": 10000,
      },
    ],
  },
  "Q:Dynamic2________["A"]": {
    "count": 5,
    "data": [
      {
        "_id": "0000000000000000000000a1",
        "routes": [
          "A",
          "B",
        ],
      },
      {
        "_id": "0000000000000000000000a2",
        "routes": [
          "A",
          "C",
        ],
      },
      {
        "_id": "0000000000000000000000b2",
        "routes": [
          "A",
          "D",
        ],
      },
      {
        "_id": "0000000000000000000000b4",
        "routes": [
          "A",
          "B",
          "C",
        ],
      },
      {
        "_id": "0000000000000000000000c2",
        "routes": [
          "A",
        ],
      },
    ],
  },
  "Q:Dynamic3________[{"$oid":"0000000000000000000000d2"}]": {
    "count": 0,
    "data": [],
  },
  "Q:Populated1______[]": {
    "count": 11,
    "data": [
      {
        "_id": "0000000000000000000000a1",
      },
      {
        "_id": "0000000000000000000000a2",
        "driver": {
          "__v": 0,
          "_id": "0000000000000000000000d1",
          "name": "Henry",
          "salary": 55000,
        },
      },
      {
        "_id": "0000000000000000000000a3",
        "driver": {
          "__v": 0,
          "_id": "0000000000000000000000d2",
          "name": "John",
          "salary": 61000,
        },
      },
      {
        "_id": "0000000000000000000000a4",
        "driver": {
          "__v": 0,
          "_id": "0000000000000000000000d3",
          "name": "Jane",
          "salary": 52000,
        },
      },
      {
        "_id": "0000000000000000000000b1",
      },
      {
        "_id": "0000000000000000000000b2",
        "driver": {
          "__v": 0,
          "_id": "0000000000000000000000d1",
          "name": "Henry",
          "salary": 55000,
        },
      },
      {
        "_id": "0000000000000000000000b3",
        "driver": {
          "__v": 0,
          "_id": "0000000000000000000000d2",
          "name": "John",
          "salary": 61000,
        },
      },
      {
        "_id": "0000000000000000000000b4",
        "driver": {
          "__v": 0,
          "_id": "0000000000000000000000d3",
          "name": "Jane",
          "salary": 52000,
        },
      },
      {
        "_id": "0000000000000000000000c1",
      },
      {
        "_id": "0000000000000000000000c2",
      },
      {
        "_id": "0000000000000000000000c3",
        "driver": {
          "__v": 0,
          "_id": "0000000000000000000000d4",
          "name": "Mary",
          "salary": 72000,
        },
      },
    ],
  },
  "Q:Populated2______[]": {
    "count": 11,
    "data": [
      {
        "_id": "0000000000000000000000a1",
      },
      {
        "_id": "0000000000000000000000a2",
        "driver": {
          "_id": "0000000000000000000000d1",
          "name": "Henry",
        },
      },
      {
        "_id": "0000000000000000000000a3",
        "driver": {
          "_id": "0000000000000000000000d2",
          "name": "John",
        },
      },
      {
        "_id": "0000000000000000000000a4",
        "driver": {
          "_id": "0000000000000000000000d3",
          "name": "Jane",
        },
      },
      {
        "_id": "0000000000000000000000b1",
      },
      {
        "_id": "0000000000000000000000b2",
        "driver": {
          "_id": "0000000000000000000000d1",
          "name": "Henry",
        },
      },
      {
        "_id": "0000000000000000000000b3",
        "driver": {
          "_id": "0000000000000000000000d2",
          "name": "John",
        },
      },
      {
        "_id": "0000000000000000000000b4",
        "driver": {
          "_id": "0000000000000000000000d3",
          "name": "Jane",
        },
      },
      {
        "_id": "0000000000000000000000c1",
      },
      {
        "_id": "0000000000000000000000c2",
      },
      {
        "_id": "0000000000000000000000c3",
        "driver": {
          "_id": "0000000000000000000000d4",
          "name": "Mary",
        },
      },
    ],
  },
  "Q:Sorted1_________[]": {
    "count": 5,
    "data": [
      {
        "_id": "0000000000000000000000b3",
      },
      {
        "_id": "0000000000000000000000b4",
      },
      {
        "_id": "0000000000000000000000c1",
      },
      {
        "_id": "0000000000000000000000c2",
      },
      {
        "_id": "0000000000000000000000c3",
      },
    ],
  },
  "Q:Static1_________[]": {
    "count": 4,
    "data": [
      {
        "_id": "0000000000000000000000a1",
      },
      {
        "_id": "0000000000000000000000a2",
      },
      {
        "_id": "0000000000000000000000a3",
      },
      {
        "_id": "0000000000000000000000a4",
      },
    ],
  },
  "Q:Static2_________[]": {
    "count": 5,
    "data": [
      {
        "_id": "0000000000000000000000b3",
        "price": 8000,
      },
      {
        "_id": "0000000000000000000000b4",
        "price": 10000,
      },
      {
        "_id": "0000000000000000000000c1",
        "price": 20000,
      },
      {
        "_id": "0000000000000000000000c2",
        "price": 40000,
      },
      {
        "_id": "0000000000000000000000c3",
        "price": 60000,
      },
    ],
  },
  "Q:Targeted1_______[[{"$oid":"0000000000000000000000b1"},{"$oid":"0000000000000000000000b2"}]]": {
    "count": 2,
    "data": [
      {
        "__v": 0,
        "_id": "0000000000000000000000b1",
        "kind": "truck",
        "price": 2000,
        "routes": [],
      },
      {
        "__v": 0,
        "_id": "0000000000000000000000b2",
        "kind": "truck",
        "price": 4000,
        "routes": [
          "A",
          "D",
        ],
      },
    ],
  },
  "Q:Unique1_________[{"$oid":"0000000000000000000000a4"}]": {
    "count": 1,
    "data": [
      {
        "__v": 0,
        "_id": "0000000000000000000000a4",
        "driver": "0000000000000000000000d3",
        "kind": "car",
        "price": 4000,
        "routes": [
          "B",
          "C",
        ],
      },
    ],
  },
}
`;
